src/library/scala/collection/immutable/Seq.scala

```diff
@@ -52,7 +52,7 @@ trait IndexedSeq[+A] extends Seq[A]
           // but if apply is more efficient than Iterators then we can use the apply for all the comparison
           // we default to the minimum preferred length
           val maxApplyCompare = {
-            val preferredLength = Math.min(applyPreferredMaxLength, that.applyPreferredMaxLength).toLong
+            val preferredLength = Math.min(applyPreferredMaxLength, that.applyPreferredMaxLength)
             if (preferredLength <= (length << 1)) Math.min(preferredLength, length) else length
```

Comment:
> That `.toLong` looks to me like it might be there on purpose in order to avoid boundary/overflow conditions around `Int.MaxValue`...?

If this was the intention, the code looks wrong to me anyway. Let's see what this conversion implies:
- With this conversion `maxApplyCompare` becomes a `Long`. However, it shouldn't matter since its value can't be greater than `length` which is a `Int`
- What happens if `length << 1` overflows:
  ```scala
  scala> 1L <= (Int.MaxValue << 1)
  res0: Boolean = false
  ```
  Ouch! It doesn't matter if `preferredLength` is a `Long`. On the other hand:
  ```scala
  scala> 1L <= (Int.MaxValue.toLong << 1)
  res0: Boolean = true
  ``` 
  I am happy to rewrite this condition to `preferredLength.toLong <= (length.toLong << 1)` if this is the intended behaviour

Reply:
I doubt that this was the intended behavior, it doesn't make much sense to me. Why doesn't it always use `Math.min(preferredLength, length)`?  Maybe the intention is: If the length is up to twice the `preferredLength`, use `apply` for the whole sequence anyway because exceeding the `preferredLength` is probably still better than creating the Iterators for only a few elements:

```
if (length > (preferredLength << 1)) preferredLength else length
```

Reply:
@allanrenucci still want to get this across the finish line?

Reply:
I've updated the PR addressing the last comment a while ago. One slight difference compared to the suggested change: I added a conversion to `Long` to prevent overflow while shifting:
```diff
- if (length > (preferredLength << 1)) preferredLength else length
+ if (length > (preferredLength.toLong << 1)) preferredLength else length
```

Reply:
ok, LGTM, @szeiger you want to hit "merge"?